### ISSUE TITLE
Highlight ellipsis in some way?

### ISSUE THREAD

**Issue by antoine-levitt** [BioTurboNick/AbbreviatedStackTraces.jl#11]
Ellipsis are quite easy to miss in a stacktrace, and should be highlighted better. Maybe by putting them at the first column, so they stand out from the [n]?

**Comment by BioTurboNick** [comment]
Worth considering. Though I kind of like the visual alignment when it's a gap in the middle of a trace, and the numbers jumping are another clue. If someone is studying the trace closely, would it really be missed?

But I won't let my personal aesthetics stand in the way if it turns out most don't like it as is :-)

**Comment by antoine-levitt** [comment]
> Worth considering. Though I kind of like the visual alignment when it's a gap in the middle of a trace, and the numbers jumping are another clue. If someone is studying the trace closely, would it really be missed?

For me the point of doing this is to *not* study the trace closely :-p having a clearer visual demarcation makes it easier to parse what's going on at a glance. As it stands, the two-line plus sometimes ellipsis pattern is hard to get. This would possibly be different with one-line frames (without signature)

**Comment by BioTurboNick** [comment]
Could just add the missing frame range.  Would stand out more and be easier to see because it is consistent?

![image](https://user-images.githubusercontent.com/1438610/115995132-057b7d80-a5a8-11eb-80ae-8cf42fa4e24a.png)

Could also remove the top or even both ellipses.

![image](https://user-images.githubusercontent.com/1438610/115995153-13c99980-a5a8-11eb-8a92-22af17e62305.png)

![image](https://user-images.githubusercontent.com/1438610/115995166-20e68880-a5a8-11eb-919f-b0a19c736241.png)

Or put the ellipsis inside

![image](https://user-images.githubusercontent.com/1438610/115995255-8aff2d80-a5a8-11eb-8138-add65cfedefd.png)

**Comment by BioTurboNick** [comment]
I kind of like the last option. Makes it stand out a little more, stands in for the function name.

Although it does interrupt the two-lines-per-frame flow.

Maybe this? The one-line version could be used in the removed-type-info version.
![image](https://user-images.githubusercontent.com/1438610/115995713-52f8ea00-a5aa-11eb-8c23-c3a61a114671.png)

Or for more consistency:
![image](https://user-images.githubusercontent.com/1438610/115995794-b420bd80-a5aa-11eb-8de4-f457ed5965dd.png)

**Comment by antoine-levitt** [comment]
How about
```
[ 9-]  Module1: function1
[-17] Module2: function2
```
(- to be possibly replaced by some fancy unicode character)?

**Comment by antoine-levitt** [comment]
Rationale is that you don't care so much about the modules that have gone between them, but this is a concise way to display how you got in and out. The two-lines [] doesn't break the pattern too much while still clarifying that some things have been elided.

**Comment by antoine-levitt** [comment]
Also the one-line [] version needs quite a bit more horizontal space

**Comment by BioTurboNick** [comment]
I don't know if "how you got out" is that useful because that's an internal detail of the omitted code. Unless you're familiar with the internals of those packages, it won't tell you much, especially without file and line information.

I could see the merit in just showing the module for the next frame, but then that's not too different from just showing the next frame, which was something you didn't like in #10 . Not clear what the right behavior there would be if it was inlined or a macro expansion, with no module attached.

**Comment by BioTurboNick** [comment]
IMO it needs to be very clear that there's "stuff we're not showing", and listing one or two modules definitively but not all suggests to me there are no others to see. I think the omission indicator probably needs to either summarize what was omitted in a high-level way, or show no indication of what was omitted, but not in between.

**Comment by mcabbott** [comment]
The last one "for more consistency" seems best IMO. Although perhaps is may as well spell it out, like `[9-17] hidden` or something, compare `[19] top-level scope`. 

Or even, since the space is free, something like:
<tt>
  [9-17] functions **error**, **apply_recipe**, **_process_seriesrecipe**, **recipe_pipeline!**, ... 
</tt>
<tt>
    @ Base, RecipesBase, Plots
</tt>

(I think the modules also want some commas, in any scheme.)

**Comment by antoine-levitt** [comment]
> I don't know if "how you got out" is that useful because that's an internal detail of the omitted code. Unless you're familiar with the internals of those packages, it won't tell you much, especially without file and line information.

OK that makes sense, I was trying to come up with examples but if something calls back into your code your function is probably something like `my_callback` and so it should be clear where that call comes from even without the frame.

> I could see the merit in just showing the module for the next frame, but then that's not too different from just showing the next frame, which was something you didn't like in #10 

My problem with that is that it looked too much like a "normal frame" and it wasn't very clear why this one should be displayed but the others omitted if you didn't know the logic.

> I think the omission indicator probably needs to either summarize what was omitted in a high-level way, or show no indication of what was omitted, but not in between.

OK I agree with this.

> Or even, since the space is free, something like

Wouldn't say space is free, it would be good to have it fit in one line.

So how about
```
[9-17] Hidden stacktrace, starting from Plots.plot
         @ Base, RecipesBase, Plots 
```
? It does "waste" two characters of space for everybody which I'm not a huge fan of, but maybe that's not too bad...

**Comment by BioTurboNick** [comment]
I hear what you mean about wanting to call out that next frame as different... I'm not opposed, but I think for the moment I'll just leave those all in. I'll think about it more though.

Added commas to the modules lists, good idea. Implemented.

Verbiage for the hidden line may be good... I'll think about that a bit.

Re: listing the functions omitted, my opinion is that the intention should be to avoid showing the user the messy internals unless they request it. But if it turns out others think that would be useful to see up front, we could revisit that?

**Comment by BioTurboNick** [comment]
Figured out I could make the ellipsis bold, so that will help a bit too.

**Comment by BioTurboNick** [comment]
How about this?
```
   [1-5] ⋮ [internal frames]
       @ Base, RecipesPipeline, [inlined methods]
```
The intended logic, as far as I've been concerned, is that the omitted frames are internals. So... let's just say they're internal?

Hidden might be okay, though seemed redundant because it's clear they're hidden, but what's missing is the why I think?

**Comment by antoine-levitt** [comment]
Not sure the [inlined methods] is helpful? I also don't like "internal" because what's internal and not depends very much on your point of view. "skipped"?

**Comment by BioTurboNick** [comment]
> Not sure the [inlined methods] is helpful? 

Yeah it's not great... the problem is that inlined methods can come from any module, so you could otherwise be missing something from the module list.

I suppose we could try to infer the module name from the path, but that could be fragile?

> I also don't like "internal" because what's internal and not depends very much on your point of view. "skipped"?

I'd argue the point of view is determined by the user, by what they have `dev`ed? If there's resistance to that more generally I'm willing to change it.

**Comment by antoine-levitt** [comment]
> Yeah it's not great... the problem is that inlined methods can come from any module, so you could otherwise be missing something from the module list.

The solution is to track the origin module for the inline functions, that should probably be a separate issue

**Comment by mcabbott** [comment]
I have a few gripes about the latest iteration:
* `⋮ [internal frames]` seems odd: Why the dots if `[1-7]` is clear? It could be `[1...7]` if you feel dots are necessary. Why re-use square brackets? `top-level scope` is the relevant prior art. Why re-use the colour which indicates paths for a comment? And why "internal", as mentioned, the relevant fact is that they are hidden. (But of course this is a minor printing thing!)
* More importantly, it seems happy to remove one stack frame, and replace it with something no shorter but more confusing. Can this be avoided? Is it easy to demand (say) at least 3 missing frames?

**Comment by antoine-levitt** [comment]
> I'd argue the point of view is determined by the user, by what they have deved? If there's resistance to that more generally I'm willing to change it.

The thing is that "internal" suggests "don't look here", which is not really what we want to convey here: it's more "I'm hiding this for your convenience, but they're here if you want to look at them". Skip conveys this better I think

**Comment by antoine-levitt** [comment]
> More importantly, it seems happy to remove one stack frame, and replace it with something no shorter but more confusing. Can this be avoided? Is it easy to demand (say) at least 3 missing frames?

This goes with the fact that we print an extra frame. If we don't print that extra frame this is less of an issue.

**Comment by mcabbott** [comment]
What does "we print an extra frame" mean? 

I hit some examples where instead of [1] to [8] all having the call information, [1] and [7] were this mysterious internal thing, while still printing [1] to [8]. My proposal is that this should never happen --- replacing the full information with a message of how to get it, without saving space, is a loss. Only once it's replacing say [1-3] should a change happen. But I don't know whether this is easy to achieve.

(You could also take the point of view that the idea is to hide internal details, not primarily to save space. In which case this could make sense.)

**Comment by antoine-levitt** [comment]
> What does "we print an extra frame" mean?

See https://github.com/BioTurboNick/AbbreviatedStackTraces.jl/issues/10: we omit all the frames in the hidden region except one (the one that entered the region). So your "[1] internal" is in fact two "internal" frames, one of which is printed but not the other. I find this not easy to see if you don't know it's there. If we don't print that one frame, you'd get "[1-2] internal".

**Comment by mcabbott** [comment]
What was happening was a middle number `[7] fun(x::` was replaced with `[7] [internal]`. Although I did not save this example... but like [41] here https://github.com/BioTurboNick/AbbreviatedStackTraces.jl/issues/10#issuecomment-826265947 with different printing. Omitting just one frame, not a sequence, seems like something to avoid.

**Comment by BioTurboNick** [comment]
> (You could also take the point of view that the idea is to hide internal details, not primarily to save space. In which case this could make sense.)

Yes, that's where I'm coming from. I see those two missions as related. The sheer number of frames and the exposure of internal functions both get in the way, especially for new users.

But I take your point that if there happen to be traces that only have one-off frames skipped, it doesn't save any space. Is that common though?

~~But @mcabbott it sounds like you're describing a situation where a stack frame is printed twice as a singular one and also as part of a range, I think? That would be a bug if so.~~ Nevermind, I understand now.

**Comment by BioTurboNick** [comment]
> * `⋮ [internal frames]` seems odd: Why the dots if `[1-7]` is clear? It could be `[1...7]` if you feel dots are necessary. Why re-use square brackets? `top-level scope` is the relevant prior art. Why re-use the colour which indicates paths for a comment? And why "internal", as mentioned, the relevant fact is that they are hidden. (But of course this is a minor printing thing!)

Good question. If I'm understanding right, `top-level scope` is the name of a section of code, as is `macro expansion`. Both are names of containers for the following frames, yes? Does it make sense to follow that for a collapsing of multiple frames? I could be persuaded, but we want these skipped frames to stand out as much as possible, don't we?

As it stands, I see the ellipses as supplanting the function name, and `[internal frames]` as the explanation for what's happening, without distracting from the bold-white function names of the present frames. (could just say `[internal]`, or get rid of the brackets though; I was following the `[inlined]` convention),

**Comment by mcabbott** [comment]
Sorry, that's quite a grumpy list! 

But I would vote to keep it simple, like `[1-7] hidden` or `[9-14] not shown` or something (if not printing other useful information). If this isn't bold, it won't look like a function name. (I guess making "macro expansion" bold is a bug?) If it always has a dash, that helps distinguish it too. "[inlined]" has brackets to separate it from the file name it appears after.

**Comment by BioTurboNick** [comment]
No worries! I'll try it out that way. I think I'm stuck on the ellipsis because I had it originally and don't want to see it go haha.

**Comment by BioTurboNick** [comment]
I'm not sure what I think about the ellipses being gone. If I'm reading down the function names in the minimal version it reads to me like a "skip" and IMO draws my eye down to the next one. That doesn't happen with the main version though. On the other hand, the bold white function names stand out more at a glance when it's gone.

![image](https://user-images.githubusercontent.com/1438610/116326810-5e5b3980-a793-11eb-96ff-5ec8007a59ec.png)

**Comment by BioTurboNick** [comment]
I think I'm happy with the format as it is, so I'm going to close this. Any future discussion can be done if/when Julia Base may integrate it :-)
